### PR TITLE
S2 ard s3 sync

### PR DESCRIPTION
**Pull request by kieranricardo**
This PR adds the `nci_s2_upload_s3_v2`, `k8s_index_s2_nbart_odc`, and `k8s_index_s2_nbart_sandbox` with the new aws users, queues, and sns defined in the current Datakube PRs. Once those PRs are merged an AWS airflow conneciton needs to be created for the S2 upload user  with the connection id `dea_public_data_sentinel2_ard`

### CODE REVIEW

**Comment by alexgleith** [comment]
Linting is failing :-)

File: dags/sentinel_2_nbart/k8s_index_s2_nbart_odc.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+"""
+# Sentinel-3 indexing automation
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/k8s_index_s2_nbart_odc.py:2]: Sentinel-2

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+from eodatasets3.stac import to_stac_item
+from eodatasets3.model import (
+    DatasetDoc,
+    ProductDoc,
+    AccessoryDoc,
+    Location,
+)
+
+import rasterio
+from eodatasets3 import serialise, validate, images, documents
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:34]: Can we run some kind of automatic sorting of imports please. There's some duplication and confusing ordering.

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+from rasterio import DatasetReader
+from shapely.geometry.polygon import Polygon
+
+NCI_DIR = '/g/data/if87/datacube/002/S2_MSI_ARD/packaged'
+S3_BUCKET = 'dea-public-data'
+WORK_DIR = Path('/g/data/v10/work/s2_nbart_rolling_archive')
+SNS_ARN = "arn:aws:sns:ap-southeast-2:060378307146:dea-public-data-sentinel-2-ard"
+
+S3 = None
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:43]: ?

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            _LOG.info(f"Completed uploaded: {future.result()}")
+
+
+def upload_dataset(granule_id):
+    """
+    :param granule_id: the id of the granule in format 'date/tile_id'
+    """
+    session = boto3.session.Session()
+
+    if not check_granule_uploaded(granule_id, session):
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:98]: I would love if we were consistent with dataset/granule/scene. Let's pick one and just use that! My preference is `dataset`.

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+@click.option('--workers', type=int, default=10)
+def main(granule_ids, workers):
+    """
+    Script to sync Sentinel-2 data from NCI to AWS S3 bucket
+    Pass in a file containing destination S3 urls that need to be uploaded.
+    """
+    setup_logging()
+
+    global S3
+    S3 = s3_client()
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:81]: Don't like the global client...

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    eo3 = create_eo3(local_path)
+    stac = to_stac_item(
+        eo3,
+        stac_item_destination_url=s3_path_stac,
+        odc_dataset_metadata_url=s3_path_eo3,
+        dataset_location=s3_path,
+    )
+    stac = json.dumps(stac, default=json_fallback, indent=4)
+
+    s3_dump(yaml.safe_dump(serialise.to_doc(eo3), default_flow_style=False), s3_path_eo3, S3)
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:183]: This doesn't need an S3 object, it'll make its own if we leave it out.

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+    Sends stac metadata as an SNS message
+
+    :param stac: serialized stac metadata
+    :param session: boto3 Session object
+    """
+
+    resource = session.client('sns', region_name='ap-southeast-2')
+    resource.publish(
+        TopicArn=SNS_ARN,
+        Message=json.dumps({'default': stac}),
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:199]: Why is this in the `default` key?

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+        dataset_location=s3_path,
+    )
+    stac = json.dumps(stac, default=json_fallback, indent=4)
+
+    s3_dump(yaml.safe_dump(serialise.to_doc(eo3), default_flow_style=False), s3_path_eo3, S3)
+
+    return stac, s3_path_stac
+
+
+def send_stac_sns(stac, session):
```
Comment (alexgleith) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:188]: If we're pushing out STAC, we should be attaching attributes to the notification too.

Check this: https://github.com/GeoscienceAustralia/dea-airflow/blob/master/scripts/c3_to_s3_rolling.py#L155

**Review by alexgleith — CHANGES_REQUESTED** [review]
A bunch of comments.

Can we please look at the C3 to S3 rolling script and share functionality as much as we can.

https://github.com/GeoscienceAustralia/dea-airflow/blob/master/scripts/c3_to_s3_rolling.py

File: dags/sentinel_2_nbart/upload_s2_nbart.py (hunk 4 above)
Comment (kieranricardo) [reply dags/sentinel_2_nbart/upload_s2_nbart.py:94]: the `c3_to_s3_rolling..` uses the granule terminology, so I'll use granule here as well to make it consistent when I reuse some of the `c3_to_s3_rolling..` code

Comment (alexgleith) [reply dags/sentinel_2_nbart/upload_s2_nbart.py:94]: I actually prefer scene. My heirarchy of naming preferences: scene > dataset > granule...

But even more important is consistency!

**Comment by kieranricardo** [comment]
hey @alexgleith, I think I responded to all of your review comments. Re. sharing code, I changed some of the `c3_rolling...` functions to be more general and import them into my code. I also added an extra task in the dag to upload the `c3_rolling..` script so that its functions can be used on the NCI.

fyi this seems to be breaking CI bc `upload_s2_nbart.py` is not in the same directory as `c3_rolling` (but they are in the same directory when they're uploaded on NCI so the code works). One fix would be to move the `upload_s2_nbart.py` file into scripts, but I personally like all the related files and DAGs being in the same folder. Whadya reckon?

**Comment by kieranricardo** [comment]
@alexgleith I've got CI to pass now :)

File: scripts/upload_s2_nbart.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        eo3.geometry = Polygon(coords[0])
+
+    for measurement in eo3.measurements.values():
+        if measurement.grid is None:
+            measurement.grid = 'default'
+
+    return eo3
+
+
+def note_measurement(
```
Comment (alexgleith) [inline scripts/upload_s2_nbart.py:276]: This is copied from EODatasets, right? Can you remind me why?

**Review by alexgleith — CHANGES_REQUESTED** [review]
Is that function from EO Dataset definitely needed in here?

File: scripts/upload_s2_nbart.py (hunk 9 above)
Comment (kieranricardo) [reply scripts/upload_s2_nbart.py:276]: This optimises `note_measurement` to not read in all the tiff files if they're not being used to create the valid data polygon. In our case this speeds things up 20x! (bc file reads on the NCI are so slow)

I also put this into a eodatasets  PR (https://github.com/GeoscienceAustralia/eo-datasets/pull/129) but AFAIK they haven't done another release

Comment (alexgleith) [reply scripts/upload_s2_nbart.py:276]: Ok, cool.

Let's hassle Jeremy for a release!

Comment (kieranricardo) [reply scripts/upload_s2_nbart.py]: @alexgleith `deamodule` built and is now using the latest release of `eodatasets3`! I changed the code here to used the `eodatasets3` `.note_measurement` method and deleted mine :)

**Review by alexgleith — APPROVED** [review]

**Pull request merged**